proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,250 @@
+class flag accordingly. Otherwise, we know the decl is not a resource and not basic, and so
+the `other` flag is set.<br>
+If the decl is not implicit, then we check if it is a struct or class type, and if so, the
+decl is a UDT, so the `udt` flag is set. We then recurse through the members of the UDT and
+collect information on what types of resources are contained by the UDT (uav, cbv, srv, or
+sampler). The corresponding resouce class flags are set. If the decl is not a struct or class
+type, then it is a basic type, and so the `basic` flag is set.<br>
+The last step is to simply check if the decl is in the `$Globals` scope. If so, then the
+`default_globals` flag will be set.<br><br>
+
+Now enough information has been gathered to determine the right diagnostics to emit, if any.
+Firstly, if `other` is set, then the variable type is not valid for the given register type.
+So, we can emit this diagnostic: `error: '%0' is an invalid resource type for register type '%1'`
+If the `resource` flag is set and the `udt` flag is not set, then we have to check the 
+register statement, and the resource class flag, and verify that the register type matches the 
+resource class flag. If not, we emit this diagnostic: 
+`error: %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used.`
+If the `basic` flag is set, then we first check if `default_globals` is set. If so, then 
+we check the register type. If the register type is 'i' or 'b', the deprecated warning
+that is treated as an error as shown in the examples above will be emitted. If 'c' is given,
+no errors will be emitted. If `default_globals` is not set and 'c' is given, a warning will
+be emitted that is treated as an error by default: `warning: register binding 'c' should only be used in global contexts. Disable with --Wno-disallow-legacy-binding-rules` 
+After this point, `default_globals` doesn't need to be set.
+If 't', 'u', or 's' are given, then this diagnostic will be emitted:
+`error: '%0' is an invalid resource type for register type '%1'`. If any other register type
+is seen, this error will be emitted:
+`error: register binding type '%1' not supported for variable of type '%0'`<br><br>
+
+Finally, in the case that `udt` is set, we first check `default_globals`. If it is set,
+then we can permit the 'c' register type if the UDT has at least one numeric member. If
+there is no numeric member, a warning that is treated by default as an error will be 
+emitted: `"warning: 'c' register type should only be used for basic types. Disable with --Wno-disallow-legacy-binding-rules"`
+Otherwise, cbuffers and tbuffers are not permitted within UDT's, and so we don't need to
+check if 'c' is given within the UDT nor emit a warning for legacy behavior that is 
+treated as an error. An error will already be emitted preventing the declaration of 
+these buffers within a UDT. After this point, `default_globals` doesn't need to be set.

```

Comment:
This is confusing, and I don't think it's correct.  And no need to talk about something that's just invalid syntax (trying to define a cbuffer or tbuffer block inside a struct or class definition).

```suggestion
Finally, handle the `udt` case.

For register type 'c': if `default_globals` is not set, we emit the same default-error
warning as for `basic` types; otherwise, if `numeric` is not set, we emit a warning
that is treated as an error by default:
`warning: register 'c' used on type with no contents to allocate in a constant buffer`

```